python_modules/dagster/dagster_tests/definitions_tests/auto_materialize_tests/asset_daemon_scenario.py

```diff
@@ -0,0 +1,246 @@
+        )
+
+    def resolve(self) -> Tuple[AutoMaterializeRuleEvaluation, Optional[Sequence[str]]]:
+        return (
+            AutoMaterializeRuleEvaluation(
+                rule_snapshot=self.rule.to_snapshot(),
+                evaluation_data=self.rule_evaluation_data,
+            ),
+            self.partitions,
+        )
+
+
+class AssetDaemonScenarioState(NamedTuple):
+    asset_specs: Sequence[AssetSpec]
+    current_time: datetime.datetime = pendulum.now()
+    run_requests: Sequence[RunRequest] = []
+    cursor: AssetDaemonCursor = AssetDaemonCursor.empty()
+    evaluations: Sequence[AutoMaterializeAssetEvaluation] = []
+    logger: logging.Logger = logging.getLogger("dagster.amp")
+    internal_instance: Optional[DagsterInstance] = None
+
+    @property
+    def instance(self) -> DagsterInstance:
+        return check.not_none(self.internal_instance)
+
+    @property
+    def assets(self) -> Sequence[AssetsDefinition]:
+        def fn() -> None:
+            ...
+
+        assets = []
+        params = {"key", "deps", "group_name", "code_version", "auto_materialize_policy"}
+        for spec in self.asset_specs:
+            assets.append(
+                asset(compute_fn=fn, **{k: v for k, v in spec._asdict().items() if k in params})
+            )
+        return assets
+
+    @property
+    def asset_graph(self) -> AssetGraph:
+        return AssetGraph.from_assets(self.assets)
+
+    def with_asset_properties(
+        self, keys: Optional[Iterable[CoercibleToAssetKey]] = None, **kwargs
+    ) -> "AssetDaemonScenarioState":
+        """Convenience method to update the properties of one or more assets in the scenario state."""
+        new_asset_specs = []
+        for spec in self.asset_specs:
+            if keys is None or spec.key in {AssetKey.from_coercible(key) for key in keys}:
+                new_asset_specs.append(spec._replace(**kwargs))
+            else:
+                new_asset_specs.append(spec)
+        return self._replace(asset_specs=new_asset_specs)
+
+    def with_all_eager(self) -> "AssetDaemonScenarioState":
+        return self.with_asset_properties(auto_materialize_policy=AutoMaterializePolicy.eager())
+
+    def with_current_time(self, time: str) -> "AssetDaemonScenarioState":
+        return self._replace(current_time=pendulum.parse(time))
+
+    def with_runs(self, *run_requests: RunRequest) -> "AssetDaemonScenarioState":
+        with pendulum.test(self.current_time):
+            for run_request in run_requests:
+                materialize(
+                    assets=self.assets,
+                    instance=self.instance,
+                    partition_key=run_request.partition_key,
+                    tags=run_request.tags,
+                    raise_on_error=False,
+                    selection=run_request.asset_selection,
+                )
+        return self
+
+    def add_requested_runs(self) -> "AssetDaemonScenarioState":
```

Comment:
Are there any scenarios where we don't want to auto-add requested runs to the tick?

Or is this just to simplify the storage model

Reply:
also, do we need to set `self.run_requests` to be a null value after this?

So `add_requested_runs` is idempotent

Reply:
Reasonable Q's -- re: cases where we don't want to auto-add requested runs, it's all based around the fact that "adding a run" right now means having a run successfully complete. So you might want to test cases where no run has started yet by the next tick, or it's started but not completed, or it's failed in the middle, all of which could have different results.

I think we'd want to keep `self.run_requests` as is -- you're right that this causes idempotence issues, but right now self.run_requests is meant to represent "runs requested by the previous evaluation", so if this nulled out run_requests, then you'd get different behavior depending on if you did `evaluate_tick().assert_run_requests().add_requested_runs()` vs `evaluate_tick().add_requested_runs().assert_run_requests()`

Reply:
ahhh --- got it. This all makes sense, thanks for explaining